src/java.base/share/classes/sun/nio/ch/SelectorImpl.java

```diff
@@ -135,6 +137,24 @@ private int lockAndDoSelect(Consumer<SelectionKey> action, long timeout)
         }
     }
 
+    private int lockAndDoSelect(Consumer<SelectionKey> action, long timeout)
+        throws IOException
+    {
+        // filter selectNow ops from consideration (timeout == 0)
+        if ((timeout == 0) || (!SelectorSelectEvent.enabled())) {
+            return implLockAndDoSelect(action, timeout);
+        }
+        long start = SelectorSelectEvent.timestamp();
+        int n = implLockAndDoSelect(action, timeout);
+        long duration = SelectorSelectEvent.timestamp() - start;
+        // always send event if timed out (n == 0)
+        if ((n == 0) || (SelectorSelectEvent.shouldCommit(duration))) {
+            timeout = (timeout < 0) ? 0 : timeout;
+            SelectorSelectEvent.commit(start, duration, n, timeout);
```

Comment:
The comment is a bit confusing here.  n == 0 means that no selected keys were added or consumed before timeout or wakeup.

Reply:
Is n == 0 intended to detect a spinning condition where the selector goes back into select when the event has not been handled?

In that case should we still emit an event if a timeout is present and the duration is greater than the timeout? For instance, in the HttpClient, we have a selector thread that will wake up at regular interval - we set up a timeout which is the min between the next expected timer event and 1500ms. So that could potentially fire an event every 1500ms if for instance the connection threshold is less than 1500ms and the connection is idle.

Maybe that's OK - and maybe in that case the onus is on the user to set a threshold greater than 1500ms?

Or should it be ((n == 0 && durationToMillis(duration) < timeoutToMillis(timeout)) || ...)) (duration and timeout probably are not in the same unit of time) - also if shouldCommit return false will the event actually be emitted down the road? Because if not then the ( n== 0 || ...) won't work.

Reply:
> Maybe that's OK - and maybe in that case the onus is on the user to set a threshold greater than 1500ms?

The threshold is 20ms so these timed-select ops in the HTTP client will record an event when they timeout.

Reply:
We should probably find a way to not emit the event if n == 0 and the operation was interrupted by Selector.wakeUp. Since we have another issue logged to emit a spin event, I wonder if we should only commit the event here if `n != 0`? The case where n == 0 would be handled by the spin event (added later) 
@AlanBateman what do you think?

Reply:
OK. I am a little concerned about how often this event will be fired when using the HttpClient - given that it's enabled by default. Idle connections sitting in the pool will fire it at least once per connection every 1500ms. That may not be too bad.

Reply:
Should I set the default to be fairly high (like maybe 1600ms)?  I think to be useful people will have to set the threshold to something that fits their needs anyway.

Reply:
> Should I set the default to be fairly high (like maybe 1600ms)? I think to be useful people will have to set the threshold to something that fits their needs anyway.

I wonder about the usefulness of this event if the default threshold is so high. My guess is that it's rare to change the default threshold. To your question then I suppose this comes back to how this event might be used. Maybe it's worth writing down a few possible usages and that might help to inform if the threshold should be 20ms or something else.